core/MatomoInstaller.php

```diff
@@ -0,0 +1,416 @@
+    }
+
+    /**
+     * Installation Step 3: Database Set-up
+     * @throws Exception|Zend_Db_Adapter_Exception
+     */
+    public static function createDatabaseObject($settings)
+    {
+        if (empty($settings['dbname']))
+        {
+            throw new Exception("No database name");
+        }
+
+        //Set some default value for database setup.
+        if (is_null($settings['type']))
+            $settings['type'] = Config::getInstance()->database['type'];
+
+        //!! MUST ADD Adapter::getRecommendedApater
+        if (is_null($settings['adapter']))
+            $settings['adapter'] = Adapter::getRecommendedApater();
+
+        if (is_null($settings['port']))
+        {
+            $port = Adapter::getDefaultPortForAdapter($settings['adapter']);
+            $settings['port'] = $port;
+        }
+
+        //DEFAULT TABLES_PREFIX could be define in global.ini.php
+        //(Could table be without prefix ?)
+        if (is_null($settings['tables_prefix']))
+            $settings['tables_prefix'] = "matomo_";
+
+        //Prepare db settings.
+        $dbInfos = array(
+            'host'          => $settings['dbhost'],
+            'username'      => $settings['dbusername'],
+            'password'      => $settings['dbpassword'],
+            'dbname'        => $settings['dbname'],
+            'tables_prefix' => $settings['tables_prefix'],
+            'adapter'       => $settings['adapter'],
+            'port'          => $settings['port'],
+            'schema'        => Config::getInstance()->database['schema'],
+            'type'          => $settings['type'],
+            'enable_ssl'    => false
+        );
+
+        if (($portIndex = strpos($dbInfos['host'], '/')) !== false) {
+            // unix_socket=/path/sock.n
+            $dbInfos['port'] = substr($dbInfos['host'], $portIndex);
+            $dbInfos['host'] = '';
+        } else if (($portIndex = strpos($dbInfos['host'], ':')) !== false) {
+            // host:port
+            $dbInfos['port'] = substr($dbInfos['host'], $portIndex + 1);
+            $dbInfos['host'] = substr($dbInfos['host'], 0, $portIndex);
+        }
+
+        try {
+            @Db::createDatabaseObject($dbInfos);
+        } catch (Zend_Db_Adapter_Exception $e) {
+            $db = Adapter::factory($adapter, $dbInfos, $connect = false);
+
+            // database not found, we try to create  it
+            if ($db->isErrNo($e, '1049')) {
+                $dbInfosConnectOnly = $dbInfos;
+                $dbInfosConnectOnly['dbname'] = null;
+                @Db::createDatabaseObject($dbInfosConnectOnly);
+                @DbHelper::createDatabase($dbInfos['dbname']);
+
+                /* // select the newly created database */
+                @Db::createDatabaseObject($dbInfos);
+            } else {
+                throw $e;
+            }
+        }
+
+        @DbHelper::checkDatabaseVersion();
+
+        @Db::get()->checkClientVersion();
```

Comment:
You should try to avoid using `@` so often. If something might throw an Error\Exception us try/catch instead.

Reply:
It was like that in the previous version. It seems installation process change since I did it (https://github.com/matomo-org/matomo/blob/4.x-dev/plugins/Installation/Controller.php)